cmd/influxd/inspect/report_tsi/report_tsi.go

```diff
@@ -0,0 +1,481 @@
+			"",
+			tsi1.WithPath(pth),
+			tsi1.DisableCompactions(),
+		)
+		if err := report.shardIdxs[id].Open(); err != nil {
+			return err
+		}
+		defer report.shardIdxs[id].Close()
+
+		// Initialise cardinality set to store cardinalities for this shard.
+		report.cardinalities[id] = map[string]*cardinality{}
+	}
+
+	// Calculate cardinalities of shards.
+	fn := report.cardinalityByMeasurement
+
+	// Blocks until all work done.
+	report.calculateCardinalities(fn)
+
+	// Print summary.
+	if err := report.printSummaryByMeasurement(); err != nil {
+		return err
+	}
+
+	allIDs := make([]uint64, 0, len(report.shardIdxs))
+	for id := range report.shardIdxs {
+		allIDs = append(allIDs, id)
+	}
+	sort.Slice(allIDs, func(i int, j int) bool { return allIDs[i] < allIDs[j] })
+
+	for _, id := range allIDs {
+		if err := report.printShardByMeasurement(id); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// calculateCardinalities calculates the cardinalities of the set of shard being
+// worked on concurrently. The provided function determines how cardinality is
+// calculated and broken down.
+func (report *reportTSI) calculateCardinalities(fn func(id uint64) error) error {
+	// Get list of shards to work on.
+	shardIDs := make([]uint64, 0, len(report.shardIdxs))
+	for id := range report.shardIdxs {
+		shardIDs = append(shardIDs, id)
+	}
+
+	errC := make(chan error, len(shardIDs))
+	var maxi uint32 // index of maximum shard being worked on.
+	for k := 0; k < report.concurrency; k++ {
+		go func() {
+			for {
+				i := int(atomic.AddUint32(&maxi, 1) - 1) // Get next partition to work on.
+				if i >= len(shardIDs) {
+					return // No more work.
+				}
+				errC <- fn(shardIDs[i])
+			}
+		}()
+	}
+
+	// Check for error
+	for i := 0; i < cap(errC); i++ {
+		if err := <-errC; err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// Cardinality struct and methods
+type cardinality struct {
+	name  []byte
+	short []uint32
+	set   *tsdb.SeriesIDSet
+}
+
+func (c *cardinality) add(x uint64) {
+	if c.set != nil {
+		c.set.AddNoLock(x)
+		return
+	}
+
+	c.short = append(c.short, uint32(x)) // Series IDs never get beyond 2^32
+
+	// Cheaper to store in bitmap.
+	if len(c.short) > useBitmapN {
+		c.set = tsdb.NewSeriesIDSet()
+		for i := 0; i < len(c.short); i++ {
+			c.set.AddNoLock(uint64(c.short[i]))
+		}
+		c.short = nil
+		return
+	}
+}
+
+func (c *cardinality) cardinality() int64 {
+	if c == nil || (c.short == nil && c.set == nil) {
+		return 0
+	}
+
+	if c.short != nil {
+		return int64(len(c.short))
+	}
+	return int64(c.set.Cardinality())
+}
+
+type cardinalities []*cardinality
+
+func (a cardinalities) Len() int           { return len(a) }
+func (a cardinalities) Less(i, j int) bool { return a[i].cardinality() < a[j].cardinality() }
+func (a cardinalities) Swap(i, j int)      { a[i], a[j] = a[j], a[i] }
+
+func (report *reportTSI) cardinalityByMeasurement(shardID uint64) error {
+	idx := report.shardIdxs[shardID]
```

Comment:
Would `Open()` + `defer Close()` the index in this func instead of the loop above

Reply:
So I tried this a few different ways, and kept getting an "index is closing" error. My change pushed this morning works in terms of printing results, and avoid opening all indices in the loop above, but I don't think I'm actually closing them? It feels like it's not really a valid fix. Let me know what you think - I know generally how defers work, and that they are pushed/popped on a stack to be processed, but it feels like this was more difficult of a fix than I was expecting.